ansible_base/authentication/utils/authentication.py

```diff
@@ -0,0 +1,129 @@
+import logging
+from typing import Optional, Tuple
+
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import AbstractUser
+from social_core.exceptions import AuthException
+from social_core.pipeline.user import get_username
+
+from ansible_base.authentication.models import Authenticator, AuthenticatorUser
+from ansible_base.authentication.social_auth import AuthenticatorStorage, AuthenticatorStrategy
+
+logger = logging.getLogger('ansible_base.authentication.utils.authentication')
+
+
+class FakeBackend:
+    def setting(self, *args, **kwargs):
+        return ["username", "email"]
+
+
+def get_local_username(user_details: dict) -> str:
+    """
+    Converts the username provided by the backend to one that doesn't conflict with users
+    from other auth backends.
+    """
+    username = get_username(strategy=AuthenticatorStrategy(AuthenticatorStorage()), details=user_details, backend=FakeBackend())
+
+    return username['username']
+
+
+def determine_username_from_uid_social(**kwargs: dict) -> dict:
+    uid = kwargs.get('details', {}).get('username', None)
+    if not uid:
+        raise AuthException('Unable to get associated username from: {{ kwargs.get("details", None) }}')
```

Comment:
Looks like an f-string was meant to be used here.

Reply:
Good catch, this actually should have been a translated string.